File: backend/src/main/java/com/server/backend/user/UserService.java
package com.server.backend.user;

import com.server.backend.user.dto.UserResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepo userRepo;

    public ResponseEntity<List<UserResponse>> getUserList() {
        var userList = new ArrayList<UserResponse>();

        for (User user : userRepo.findAll()) {
            userList.add(getUserResponse(user));
        }

        return ResponseEntity.ok(userList);
    }

    public UserResponse getUserResponse(User user) {
        return UserResponse.builder()
                .username(user.getUsername())
                .email(user.getEmail())
                .role(user.getRole().toString())
                .build();
    }

}

File: backend/src/main/kotlin/com/server/backend/user/UserService.kt
package com.server.backend.user

import com.server.backend.auth.dto.request.LogoutRequest
import com.server.backend.user.dto.response.LogoutResponse
import com.server.backend.jwt.JwtService
import com.server.backend.user.dto.response.UserResponse
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service

@Service
class UserService(
    private val userRepo: UserRepo,
    private  val jwtService: JwtService
) {
    fun getUserList(role: UserRole? = null): ResponseEntity<List<UserResponse>> {
        val users = userRepo.findAll()
            .filter { role == null || it.role == role }
            .map(::toUserResponse)

        return ResponseEntity.ok(users)
    }

    fun getUserById(id: Long): ResponseEntity<UserResponse> {
        val user = userRepo.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(toUserResponse(user))
    }

    fun getUserByUsername(username: String): ResponseEntity<UserResponse> {
        val user = userRepo.findByUsername(username) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(toUserResponse(user))
    }

    fun getUserByEmail(email: String): ResponseEntity<UserResponse> {
        val user = userRepo.findByEmail(email) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(toUserResponse(user))
    }

    fun toUserResponse(user: User) = UserResponse(
        username = user.username,
        email = user.email,
        role = user.role.toString()
    )

    fun logout(
        authHeader: String,
        request: LogoutRequest
    ) : ResponseEntity<LogoutResponse> {
        val accessToken = authHeader.substring(7)

        jwtService.blacklistToken(
            accessToken,
            jwtService.extractExpiration(accessToken)
        )

        request.refreshToken.let { refreshToken ->
            jwtService.blacklistToken(
                refreshToken,
                jwtService.extractExpiration(refreshToken)
            )
        }

        return ResponseEntity.ok(
            LogoutResponse(
                message = "Logged out successfully"
            )
        )
    }
}